rules.go

```diff
@@ -0,0 +1,237 @@
+package adblockgoparser
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+)
+
+var (
+	ErrSkipComment = errors.New("Commented rules are skipped")
+	ErrSkipHTML    = errors.New("HTML rules are skipped")
+	binaryOptions  = []string{
+		"document",
+		"domain",
+		"elemhide",
+		"font",
+		"genericblock",
+		"generichide",
+		"image",
+		"matchcase",
+		"media",
+		"object",
+		"other",
+		"ping",
+		"popup",
+		"script",
+		"stylesheet",
+		"subdocument",
+		"thirdparty",
+		"webrtc",
+		"websocket",
+		"xmlhttprequest",
+	}
+	optionsSplitPat = fmt.Sprintf(",(~?(?:%v))", strings.Join(binaryOptions, "|"))
+	optionsSplitRe  = regexp.MustCompile(optionsSplitPat)
+)
+
+// Structs
+
+type Request struct {
+	// parsed full URL of the request
+	URL *url.URL
+	// a value of Origin header
+	Origin string
+	// a value of Referer header
+	Referer string
+	// Defines is request looks like XHLHttpRequest
+	IsXHR bool
+}
+
+type Rule struct {
+	raw         string
+	ruleText    string
+	regexString string
+	regex       *regexp.Regexp
+	options     map[string]bool
+	isException bool
+	domains     map[string]bool
+}
+
+func (rule *Rule) Match(url string) bool {
+	if rule.regex == nil {
+		rule.regex = regexp.MustCompile(rule.regexString)
+	}
+	return rule.regex.MatchString(url)
+}
+
+func ParseRule(ruleText string) (*Rule, error) {
+	rule := &Rule{
+		domains:  map[string]bool{},
+		options:  map[string]bool{},
+		raw:      ruleText,
+		ruleText: strings.TrimSpace(ruleText),
+	}
+
+	isComment := strings.Contains(rule.ruleText, "!") || strings.Contains(rule.ruleText, "[Adblock")
+	if isComment {
+		return nil, ErrSkipComment
+	}
+
+	isHTMLRule := strings.Contains(rule.ruleText, "##") || strings.Contains(rule.ruleText, "#@#")
+	if isHTMLRule {
+		return nil, ErrSkipHTML
+	}
+
+	rule.isException = strings.HasPrefix(rule.ruleText, "@@")
+	if rule.isException {
+		rule.ruleText = rule.ruleText[2:]
+	}
+
+	if strings.Contains(rule.ruleText, "$") {
+		var option string
+
+		parts := strings.SplitN(rule.ruleText, "$", 2)
+		length := len(parts)
+
+		if length > 0 {
+			rule.ruleText = parts[0]
+		}
+
+		if length > 1 {
+			option = parts[1]
+		}
+
+		options := strings.Split(option, ",")
+		for _, option := range options {
+			if strings.HasPrefix(option, "domain=") {
+				rule.domains = parseDomainOption(option)
+			} else {
+				rule.options[strings.TrimPrefix(option, "~")] = !strings.HasPrefix(option, "~")
+			}
+		}
+	}
+
+	rule.regexString = ruleToRegexp(rule.ruleText)
+
+	return rule, nil
+}
+
+type RuleSet struct {
+	rules []*Rule
+}
+
+func (rules *RuleSet) Allow(url string) bool {
+	for _, rule := range rules.rules {
+		if rule.Match(url) {
```

Comment:
Unfortunately, we need to take another approach. EasyList has about 60k of rules, going through the list and matching one-by-one is quite inefficient.

One of the reasons why we decided to create this library is uneffective matching in existing ones.